### PR TITLE
FIREFLY-116:Evaluate the wavelength readouts from various types of FITS and fix any Errors

### PR DESCRIPTION
**Pull request by lznakano**
The wavelength implementation is carefully evaluated based on algorithm described in three reference papers.  Default values are enforced.  Three bugs are fixed.   Please see ticket https://jira.ipac.caltech.edu/browse/FIREFLY-116 for more information. 

Test URL: https://irsawebdev9.ipac.caltech.edu/FIREFLY-116/firefly/

### CODE REVIEW

**Comment by lznakano** [comment]
@ejoliet 
Please take a look at my comment in FIREFLY-116.

**Comment by ejoliet** [comment]
Ok, i'll take a look but just for the records here, i did tested 'test' and 'testLinear' files and wavelength is shown inn the display! Yay!! 
The one that is not has this header:
```LVF-position-20-Ch-1.fits
# HDU 0 in LVF-position-20-Ch-1.fits:
SIMPLE  =                    T / conforms to FITS standard                      
BITPIX  =                  -64 / array data type                                
NAXIS   =                    2 / number of array dimensions                     
NAXIS1  =                 2048                                                  
NAXIS2  =                 2048                                                  
WCSAXES =                    3 / Number of coordinate axes                      
CRPIX1  =               1024.0 / Pixel coordinate of reference point            
CRPIX2  =               1024.0 / Pixel coordinate of reference point            
CRPIX3  =                  0.0 / Pixel coordinate of reference point            
CDELT1  =  -0.0017222222222222 / [deg] Coordinate increment at reference point  
CDELT2  =   0.0017222222222222 / [deg] Coordinate increment at reference point  
CDELT3  =  2.7997850569788E-10 / [m] Coordinate increment at reference point    
CUNIT1  = 'deg'                / Units of coordinate increment and value        
CUNIT2  = 'deg'                / Units of coordinate increment and value        
CUNIT3  = 'm'                  / Units of coordinate increment and value        
CTYPE1  = 'RA---TAN'           / Right ascension, gnomonic projection           
CTYPE2  = 'DEC--TAN'           / Declination, gnomonic projection               
CTYPE3  = 'WAVE'               / Vacuum wavelength (linear)                     
CRVAL1  =       151.8596107642 / [deg] Coordinate value at reference point      
CRVAL2  =      1.8814331914593 / [deg] Coordinate value at reference point      
CRVAL3  =  7.4106020148532E-07 / [m] Coordinate value at reference point        
LONPOLE =                180.0 / [deg] Native longitude of celestial pole       
LATPOLE =      1.8814331914593 / [deg] Native latitude of celestial pole        
RADESYS = 'ICRS'               / Equatorial coordinate system                   
SIM-TYPE= 'LVF     '                                                            
CHAN    =                    1                                                  
PIXSCALE=                  6.2   ```

**Comment by lznakano** [comment]
@ejoliet 
I uploaded one Spherex file that I got in the confluence page.  I tested it.  As I commented in the ticket, although it seems working, the result is wrong because the header contains several wcsaxes(ka).  We need to discuss this issue if we want to handle such headers or not,  if we do want to handle it, we can do so.  I can issue another ticket to process such headers.

File: src/firefly/js/visualize/projection/WavelengthHeaderParser.js [hunk 1]
```diff
@@ -19,21 +21,147 @@ const allGoodValues= (...numbers)  => numbers.every((n) => !isNaN(n));
...
+
+    if(ctype){
+        return (nAxis+1).toString();
+    }
+    else{
+        return nAxis.toString();
+    }
+
+}
+function assignDefault(inArr, keyRoot, which, N){
```
Comment (robyww) [inline src/firefly/js/visualize/projection/WavelengthHeaderParser.js]: i would be better to name this routine more clearly.  It look like a general reusable name to it has a specific purpose.

Also you might want to document it. I am having trouble understanding what it is doing.

File: src/firefly/js/visualize/projection/WavelengthHeaderParser.js [hunk 2]
```diff
@@ -19,21 +21,147 @@ const allGoodValues= (...numbers)  => numbers.every((n) => !isNaN(n));
...
+ *    is 0 for j!=i and 1 if j==i
+ *
+ * @param header
+ * @param zeroHeader
+ * @param which
+ * @returns {string}
+ */
+function getPC_ijKey(header, zeroHeader,which){
+
+    var key;
```
Comment (robyww) [inline src/firefly/js/visualize/projection/WavelengthHeaderParser.js]: should be `let`

File: src/firefly/js/visualize/projection/WavelengthHeaderParser.js [hunk 3]
```diff
@@ -19,21 +21,147 @@ const allGoodValues= (...numbers)  => numbers.every((n) => !isNaN(n));
...
+    let hasPC=false, hasCD=false;
+    for (key in header) {
+        if (key.startsWith(`PC${which}_`)) {
+            hasPC = true;
+            break;
+        }
+    }
```
Comment (robyww) [inline src/firefly/js/visualize/projection/WavelengthHeaderParser.js]: You probably need to make a new routine in  `FitsHeaderUtil`, `makeHeaderParse` call something like `hasKeyStartingWith()` and add one to `makeDoubleHeaderParse` as well.

File: src/firefly/js/visualize/projection/WavelengthHeaderParser.js [hunk 4]
```diff
@@ -19,21 +21,147 @@ const allGoodValues= (...numbers)  => numbers.every((n) => !isNaN(n));
...
+    if (!hasCD && zeroHeader){
+        for (key in zeroHeader) {
+            if (key.startsWith(`CD${which}_`)) {
+                hasCD = true;
+                break;
+            }
+        }
+    }
+    if (hasPC && hasCD){
+        throw Error('The FITs header contains both PC_ij and CD_ij');
```
Comment (robyww) [inline src/firefly/js/visualize/projection/WavelengthHeaderParser.js]: We don't want to throw exceptions here, That has the potential to break all the new image processing. Just return a a `false` or something and return an `undefined` in the calling function.

File: src/firefly/js/visualize/projection/WavelengthHeaderParser.js [hunk 5]
```diff
@@ -43,12 +171,21 @@ function calculateWavelengthParams(parse, altWcs, wlTable) {
...
 
-    const canDoPlaneCalc= allGoodValues(crpix,crval,cdelt) && nAxis===3;
-    if (canDoPlaneCalc && (!algorithm || !pc_3j || !r_j)) {
-        return makeSimplePlaneBased(crpix, crval, cdelt, nAxis, wlType, units, 'use PLANE since is cube and parameters missing');
+
+    const canDoPlaneCalc= allGoodValues(crpix,crval,cdelt && nAxis===3 ) ;
+
+    //We only support the standard format in the FITs header.  The standard means the CTYPEka='WAVE-ccc" where
+    //ccc can be 'F2W', 'V2W', 'LOG', 'TAB' or empty that means linear. If the header does not have this kind
+    // CTYPEka defined, we check if the it is a spectra cube.  If it is a spectra cube, we display the wavelength
```
Comment (robyww) [inline src/firefly/js/visualize/projection/WavelengthHeaderParser.js]: good docs.

File: src/firefly/js/visualize/projection/WavelengthHeaderParser.js [hunk 6]
```diff
@@ -2,12 +2,14 @@
  * License information at https://github.com/Caltech-IPAC/firefly/blob/master/License.txt
  */
 
-import {makeDoubleHeaderParse} from '../FitsHeaderUtil.js';
+import {makeDoubleHeaderParse,getHeader} from '../FitsHeaderUtil.js';
```
Comment (robyww) [inline src/firefly/js/visualize/projection/WavelengthHeaderParser.js]: remove `getHeader` it is not used

File: src/firefly/js/visualize/projection/WavelengthHeaderParser.js [hunk 7]
```diff
@@ -7,7 +7,9 @@ import {AWAV, F2W, LINEAR, LOG, PLANE, TAB, V2W, WAVE} from './Wavelength.js';
 
 export function parseWavelengthHeaderInfo(header, altWcs='', zeroHeader, wlTable) {
     const parse= makeDoubleHeaderParse(header, zeroHeader, altWcs);
-    return calculateWavelengthParams(parse,altWcs, wlTable);
+    const which= altWcs?'1':getWCSAXES(parse);
+    const mijMatrixKeyRoot = getPC_ijKey(parse,which);
```
Comment (robyww) [inline src/firefly/js/visualize/projection/WavelengthHeaderParser.js]: Check and return if this value is undefined

**Pull request merged**